src/util/worker/mod.rs

```diff
@@ -145,22 +153,43 @@ pub struct Worker<T: Display> {
     scheduler: Scheduler<T>,
     receiver: Mutex<Option<Receiver<Option<T>>>>,
     handle: Option<JoinHandle<()>>,
+    timeout: Option<usize>,
+    tasks_per_tick: Option<usize>,
 }
 
-fn poll<R, T>(mut runner: R, rx: Receiver<Option<T>>, counter: Arc<AtomicUsize>, batch_size: usize)
-where
+fn poll<R, T>(
+    mut runner: R,
+    rx: Receiver<Option<T>>,
+    counter: Arc<AtomicUsize>,
+    batch_size: usize,
+    timeout: Option<usize>,
+    tasks_per_tick: Option<usize>,
+) where
     R: BatchRunnable<T> + Send + 'static,
     T: Display + Send + 'static,
 {
     let name = thread::current().name().unwrap().to_owned();
     let mut keep_going = true;
     let mut buffer = Vec::with_capacity(batch_size);
+
+    let timeout = timeout.map(|t| Duration::from_secs(t));
```

Comment:
I think we can follow the thread pool style, define a next_task function, see https://github.com/pingcap/tikv/blob/master/src/util/threadpool.rs#L276

You can use try_recv at first, then use recv or recv_timeout.